docker/pandoc-lambda/function/app.py

```diff
@@ -123,6 +127,166 @@ def promote_case_footnotes(doc):
 
 
 
+def sectionizer(doc, doc_w_twips=10080, doc_h_twips=14400, paracols=2, internal_margin_twips=720, external_margin_twips=1080, big_margin_twips=1440):
+    """
+        Adds section breaks for headers, footers, columns, etc.
+
+        doc_w doc_h:
+            document width and height in TWIPS
+            1 TWIP = 1/20th pt. (a typographical point)
+            1 inch = 1440 TWIPS
+            my local grocer is about 1300 ft. in each direction, so 1 grocery trip ~= 6.5e+7 grocery TWIPs
+
+        paracols:
+            Number of columns in non-header paragraph text. There is no way to specify # of cols in a style
+    """
+
+    doc_w = Twips(doc_w_twips)
+    doc_h = Twips(doc_h_twips)
+    internal_margin = Twips(internal_margin_twips)
+    external_margin = Twips(external_margin_twips)
+    big_margin = Twips(big_margin_twips)
+
+    # See which rIDs pandoc gave our headers and footers. These values come
+    rels = doc.part.rels
+    headers_and_footers = { rels[rid].target_ref.replace('.xml', '') : rels[rid]
+      for rid in rels if rels[rid].target_ref.endswith('.xml') and
+      # rels lists, among other things, related xml files like header and footer.
+      # target_ref property holds the file name. Check the docx word/_rels/document
+      (rels[rid].target_ref.startswith('header') or rels[rid].target_ref.startswith('footer'))
+    }
+
+    body_element = doc.element.xpath('/w:document/w:body')[0]
+
+    def section_break(destination, frontmatter=False, chapter=False, section=False, paragraph=False):
+        """ Assembles the section break, appends it to the appropriate graf, properly discards the empty wrapper,
+        because it gives a hoot and won't pollute. """
+
+        # The doc itself should have an existing default section attached to body, not a paragraph
+        if destination == 'doc-wide':
+            sec = doc.sections[0]
+        else:
+            sec = doc.add_section()
+            destination._element.xpath('w:pPr')[0].append(sec._sectPr)
+            body_element.remove(doc.paragraphs[-1]._element)
```

Comment:
I wonder if it would be faster to use a lower-level API to add the section breaks right where you want them, rather than appending to the end of the doc and then removing. Is the current behavior, which clones all the props (https://github.com/python-openxml/python-docx/blob/da75fcf01f7f322e846e2ac3e1936aedd766acc8/docx/oxml/document.py#L35) useful, or would it be more useful if you could just add a bare one where you wanted, if we could work out a nice syntax for that?

Reply:
We could do that. I use python-docx to create the section because they're surprisingly fussy to work with and doing them directly strikes me as being more fragile and difficult to maintain. For example, having the sectPr child elements in the wrong order triggers a corrupt document error, or can result in things being ignored silently. However, we are going export a lot more frequently than modify section breaks, I suppose. I was also trying to avoid using protected functions like set_sectPr() anyway, but I probably end up having to do that more in my current method, so whatever.

Creating a string version of a section break using a template (being sure to include w and r namespaces in the xml,)
parsing that, and using the appropriate paragraph's ._element.set_sectPr() function should do the trick.

Reply:
Cool beans: that's convincing. Doesn't sound worth it unless we know it to be way faster, which we don't.